nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java

```diff
@@ -0,0 +1,303 @@
+
+        TreePath enclosingMethod =
+            NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(returnState.getPath());
+        if (enclosingMethod == null) {
+          throw new RuntimeException("no enclosing method, lambda or initializer!");
+        }
+
+        // We only add returns that are directly in the method under analysis
+        if (!enclosingMethod.getLeaf().equals(methodTreeUnderAnalysis)) {
+          return super.visitReturn(node, null);
+        }
+
+        // We add the current return tree to the set as it's in the method
+        returnTreesInMethodUnderAnalysis.add(node);
+        return super.visitReturn(node, null);
+      }
+    }.scan(methodState.getPath(), null);
+  }
+
+  /*
+   * Sub-classes can only strengthen the post-condition.
+   * We check if the list in the child classes is at least the same as in the parent class.
+   */
+  @Override
+  protected void validateOverridingRules(
+      Set<String> overridingFieldNames,
+      NullAway analysis,
+      VisitorState state,
+      MethodTree tree,
+      Symbol.MethodSymbol overriddenMethod) {
+    FieldContractUtils.validateOverridingRules(
+        annotName, overridingFieldNames, analysis, state, tree, overriddenMethod);
+  }
+
+  @Override
+  public void onDataflowVisitReturn(
+      ReturnTree returnTree, NullnessStore thenStore, NullnessStore elseStore) {
+    // We only explore return statements that is inside
+    // the method under validation
+    if (!returnTreesInMethodUnderAnalysis.contains(returnTree)) {
+      return;
+    }
+
+    // We might have already found another return tree that results in
+    // what we need, so we don't keep going deep.
+    if (semanticsHold) {
+      return;
+    }
+
+    Set<String> fieldNames = getAnnotationValueArray(methodSymbolUnderAnalysis, annotName, false);
+    if (fieldNames == null) {
+      throw new RuntimeException("List of field names shouldn't be null");
+    }
+
+    boolean trueIfNonNull = getTrueIfNonNullValue(methodSymbolUnderAnalysis);
+
+    // We extract all the data-flow of the fields found by the
+    // engine in the "then" case (i.e., true case)
+    // and check whether all fields in the annotation parameter are non-null
+    Set<String> nonNullFieldsInPath =
+        thenStore.getReceiverFields(trueIfNonNull ? Nullness.NONNULL : Nullness.NULL).stream()
+            .map(e -> e.getSimpleName().toString())
+            .collect(Collectors.toSet());
+    boolean allFieldsInPathAreVerified = nonNullFieldsInPath.containsAll(fieldNames);
+
+    if (allFieldsInPathAreVerified) {
+      // If it's a literal, then, it needs to return true/false,
+      // depending on the trueIfNonNull flag
+      if (returnTree.getExpression() instanceof LiteralTree) {
+        LiteralTree expressionAsLiteral = (LiteralTree) returnTree.getExpression();
+        if (expressionAsLiteral.getValue() instanceof Boolean) {
+          this.semanticsHold = (boolean) expressionAsLiteral.getValue();
+        }
```

Comment:
I'm not sure I follow this.  I think from the comment above that if you choose the `thenStore` or `elseStore` and just check for `NONNULL` fields you shouldn't need this.  Those stores already correspond respectively to what is known when the return value is `true` or `false`, I think.

Reply:
You were indeed correct! The algorithm now picks the store based on the `result` flag!